src/verso-manual/VersoManual/Markdown.lean

```diff
@@ -208,11 +211,94 @@ def blockFromMarkdown'
+  | .nullchar => .error "Unexpected null character in parsed Markdown"
+  | .del _ => .error "Unexpected strikethrough in parsed Markdown"
+  | .em txt => joinArrM <| txt.mapM stringFromMarkdownText
+  | .strong txt => joinArrM <| txt.mapM stringFromMarkdownText
+  | .a _ _ _ txt => joinArrM <| txt.mapM stringFromMarkdownText
+  | .latexMath m => pure <| String.join m.toList
+  | .latexMathDisplay m =>  pure <| String.join m.toList
+  | .u txt => .error s!"Unexpected underline around {repr txt} in parsed Markdown:"
+  | .code strs => pure <| String.join strs.toList
+  | .entity ent => .error s!"Unsupported entity {ent} in parsed Markdown"
+  | .img .. => .error s!"Unexpected image in parsed Markdown"
+  | .wikiLink .. => .error s!"Unexpected wiki-style link in parsed Markdown"
+where
+  joinArrM (x : Except String (Array String)) : Except String String :=
+    return String.join (← x).toList
+
+open Verso.Doc.Elab
+
+/--
+Updates the active sections given a new header with `level`.
+-/
+private partial def closeSections {m} [Monad m]
+    [MonadStateOf PartElabM.State m]
+    (level : Nat) : MDT m b i Unit := do
+  let hdrs := (← getThe MDState).inHeaders
+  match hdrs with
+  | [] => modifyThe MDState ({· with inHeaders := [(level, 0)]})
+  | (docLevel, nesting) :: more =>
+    if level ≤ docLevel then
+      if let some ctxt' := (← getThe PartElabM.State).partContext.close default then -- TODO: source position!
+        modifyThe PartElabM.State fun st => {st with partContext := ctxt'}
+        closeSections level
+      if level < docLevel then
+        modifyThe MDState ({· with inHeaders := more})
+    else
+      modifyThe MDState ({· with inHeaders := (level, nesting + 1) :: hdrs})
+
+private partial def partFromMarkdownAux {m} [Monad m]
+    [MonadLiftT PartElabM m] [MonadStateOf PartElabM.State m]
+    [MonadQuotation m] [AddMessageContext m] [MonadError m]
+    : MD4Lean.Block → MDT m Term Term Unit
+  | .header level txt => do
+    closeSections level
+    let txtStxs ← txt.mapM inlineFromMarkdown |>.run' none
+    let titleTexts ← match txt.mapM stringFromMarkdownText with
+      | .ok t => pure t
+      | .error e => throwError m!"Invalid Markdown in header:\n{e}"
+    let titleText := titleTexts.foldl (· ++ ·) ""
+    PartElabM.push {
+      titleSyntax := quote (k := `str) titleText
+      expandedTitle := some (titleText, txtStxs)
+      metadata := none
+      blocks := #[]
+      priorParts := #[]
+    }
+  | b => do
+    PartElabM.addBlock (← blockFromMarkdownAux b)
+
+/--
+Adds blocks from Markdown, treating top-level headers as new parts.
+
+Note that `handleHeaders` is still used for elaborating headers that appear
+nested within blocks (e.g., blockquotes).
+
+`currentHeaderLevels` gives a list of headers within which elaboration is
+occurring and which can be terminated by the current elaboration. Typically,
+these are taken from a previous iteration of `partFromMarkdown`, but they can
```

Comment:
```suggestion
these are taken from a previous iteration of `addPartFromMarkdown`, but they can
```
?

Reply:
Yes, sorry, a casualty of an incomplete refactor. There were a couple of these, all of which I've changed in c1506a6.